fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/WhereFunction.java

```diff
@@ -0,0 +1,95 @@
+import javax.annotation.Nonnull;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+
+/**
+ * Describes a function which can scope down the previous invocation within a FHIRPath expression,
+ * based upon an expression passed in as an argument. Supports the use of `$this` to reference the
+ * element currently in scope.
+ *
+ * @author John Grimes
+ * @see <a href="https://pathling.csiro.au/docs/fhirpath/functions.html#where">where</a>
+ */
+public class WhereFunction implements NamedFunction {
+
+  private static final String NAME = "where";
+
+  @Nonnull
+  @Override
+  public FhirPath invoke(@Nonnull final NamedFunctionInput input) {
+    checkUserInput(input.getArguments().size() == 1,
+        "where function accepts one argument");
+    final FhirPath inputPath = input.getInput();
+    checkUserInput(inputPath instanceof NonLiteralPath,
+        "Input to where function cannot be a literal: " + inputPath.getExpression());
+    checkUserInput(input.getArguments().get(0) instanceof NonLiteralPath,
+        "Argument to where function cannot be a literal: " + input.getArguments().get(0)
+            .getExpression());
+    final NonLiteralPath argumentPath = (NonLiteralPath) input.getArguments().get(0);
+    checkUserInput(argumentPath instanceof BooleanPath && argumentPath.isSingular(),
+        "Argument to where function must be a singular Boolean: " + argumentPath.getExpression());
+    checkUserInput(argumentPath.getThisColumn().isPresent(),
+        "Argument to where function must be navigable from collection item (use $this): "
+            + argumentPath.getExpression());
+
+    // The result is the input value if it is equal to true, or null otherwise (signifying the
+    // absence of a value).
+    final Column argumentTrue = argumentPath.getValueColumn().equalTo(true);
+    final Column thisColumn = argumentPath.getThisColumn().get();
+
+    final Dataset<Row> dataset;
+    final Optional<Column> idColumn;
+    final List<Column> groupingColumns = input.getContext().getGroupingColumns();
+    final Dataset<Row> argumentDataset = argumentPath.getDataset();
+
+    if (!groupingColumns.isEmpty()) {
+      final Dataset<Row> distinctIds = argumentDataset
+          .select(groupingColumns.toArray(new Column[]{})).distinct();
+      final DatasetWithColumns datasetWithColumns = joinOnColumns(distinctIds, groupingColumns,
+          argumentDataset, groupingColumns, Optional.of(argumentTrue), JoinType.LEFT_OUTER);
+      dataset = datasetWithColumns.getDataset();
+      input.getContext().setGroupingColumns(datasetWithColumns.getColumns());
+      idColumn = Optional.empty();
+    } else {
+      final Column argumentId = checkPresent(argumentPath.getIdColumn());
+      final String hash = randomShortString();
+      final String idColumnName = hash + ID_COLUMN_SUFFIX;
+      Dataset<Row> distinctIds = argumentDataset.withColumn(idColumnName, argumentId);
```

Comment:
Maybe ??? 
```
Dataset<Row> distinctIds = argumentDataset.select(argumentId.alias(idColumnName)).distinct()
```

Reply:
I tried this and it fails a number of tests - I have not had much luck getting aliasing to work across joins in general.

This is why you will see a lot of `withColumn` stuff going on.

I don't completely understand why aliasing doesn't work, or why it only sometimes works.